src/cargo/core/resolver/mod.rs

```diff
@@ -298,109 +244,213 @@ fn activate(mut cx: Box<Context>,
+
+        // Filter the set of candidates based on the previously activated
+        // versions for this dependency. We can actually use a version if it
+        // precisely matches an activated version or if it is otherwise
+        // incompatible with all other activated versions. Note that we define
+        // "compatible" here in terms of the semver sense where if the left-most
+        // nonzero digit is the same they're considered compatible.
+        let my_candidates: Vec<Rc<Summary>> = candidates.iter().filter(|&b| {
+            prev_active.iter().any(|a| a == b) ||
+                prev_active.iter().all(|a| {
+                    !compatible(a.version(), b.version())
+                })
+        }).cloned().collect();
+
+        // Alright, for each candidate that's gotten this far, it meets the
+        // following requirements:
+        //
+        // 1. The version matches the dependency requirement listed for this
+        //    package
+        // 2. There are no activated versions for this package which are
+        //    semver-compatible, or there's an activated version which is
+        //    precisely equal to `candidate`.
+        //
+        // This means that we're going to attempt to activate each candidate in
+        // turn. We could possibly fail to activate each candidate, so we try
+        // each one in turn.
+        let mut remaining_candidates = RcVecIter::new(my_candidates);
+        let maybe_candidate =
+            remaining_candidates.next().map(|(_, candidate)| candidate.clone());
+        let candidate: Rc<Summary> = match maybe_candidate {
+            Some(candidate) => {
+                // We have a candidate. Add an entry to the `backtrack_stack` so
+                // we can try the next one if this one fails.
+                backtrack_stack.push(BacktrackFrame {
+                    context_backup: cx.clone(),
+                    deps_backup: remaining_deps.clone(),
+                    remaining_candidates: remaining_candidates,
+                    parent: parent.clone(),
+                    cur: cur,
+                    dep: dep.clone(),
+                    prev_active: prev_active,
+                    all_candidates: candidates,
+                });
+                candidate
+            }
+            None => {
+                // This dependency has no valid candidate. Backtrack until we
+                // find a dependency that does have a candidate to try, and try
+                // to activate that one.  This resets the `remaining_deps` to
+                // their state at the found level of the `backtrack_stack`.
+                trace!("{}[{}]>{} -- None", parent.name(), cur, dep.name());
+                let last_err = activation_error(&cx, registry, None, &parent, &dep,
+                                                &prev_active, &candidates);
+                try!(find_candidate(&mut backtrack_stack, &mut cx, &mut remaining_deps,
```

Comment:
Hm yeah this is interesting. I haven't quite groked this enough to be able to recommend a test, but stylistically it looks like this branch of the `match` should probably execute a `continue` at some point to go back up the to the top of the loop. That should end up doing all the resetting of local state as necessary, right?